R/type-vctr.R

```diff
@@ -108,9 +108,8 @@ vec_restore.vctrs_vctr <- function(x, to, ..., i = NULL) {
 #' @method vec_ptype2 vctrs_vctr
 #' @export
 vec_ptype2.vctrs_vctr <- function(x, y, ..., x_arg = "x", y_arg = "y") {
-  # This method is redundant with `vec_ptype2.default()` but it
-  # instructs `vec_c()` that it isn't a foreign type. This avoids
-  # infinite recursion through `c.vctrs_vctr()`.
+  # This method instructs `vec_c()` that it isn't a foreign type. This
+  # avoids infinite recursion through `c.vctrs_vctr()`.
   vec_default_ptype2(x, y, ..., x_arg = x_arg, y_arg = y_arg)
```

Comment:
With the new system, the code in left methods (generic-methods) is never called. The consequence is that it's now possible to implement `vec_ptype2()` methods with `vctrs_vctr`. Also it's misleading to make it seem that the code inside the left-method is called.

I wonder if it'd make sense to parameterise dispatch via literal hints:

```r
vec_ptype2.vctrs_vctr <- function(x, y, ...) {
  "vctrs:::no_ptype2"
  UseMethod("vec_ptype2.vctrs_vctr")
}

vec_ptype2.AsIs <- function(x, y, ...) {
  "vctrs:::wrapper_type"
  UseMethod("vec_ptype2.AsIs")
}
```